keps/sig-node/1985-remove-dockershim/README.md

```diff
@@ -0,0 +1,298 @@
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Terms
+
+- **CRI:** Container Runtime Interface – a plugin interface which enables kubelet to use a wide variety of container 
+runtimes, without the need to recompile.
+
+## Summary
+
+CRI for docker (i.e. dockershim) is currently a built-in container runtime in kubelet code base. This proposal aims 
+at a deprecation and subsequent removal of dockershim from kubelet.
+
+## Motivation
+
+In Kubernetes, CRI is the used as the "default" container runtime, while currently the CRI of docker (a.k.a. dockershim) 
+is part of kubelet code and coupled with kubelet's lifecycle. 
+
+This is not ideal as kubelet then has dependency on specific container runtime which leads to maintenance burden for not 
+only developers in sig-node, but also cluster administrators when critical issues (e.g. runc CVE) happen to container runtimes. The pros of removing dockershim is straightforward:
+
+### Pros
+- Docker is not special and should be just a CRI implementation just like every other CRI implementation in our ecosystem. 
+- Currently, dockershim "enjoys" some backdoors for various reasons (see [legacyLogProvider](https://cs.k8s.io/?q=legacyLogProvider&i=nope&files=&repos=kubernetes/kubernetes) for example) . Removing these "features" should eliminate maintenance burden of kubelet.
+- A cri-dockerd can be maintained independently by folks who are interested in keeping this functionality
+- Over time we can remove vendored docker dependencies in kubelet.
+
+Having said that, cons of removal built-in dockershim requires lots of attention:
+
+### Cons
+- Deployment pain with a new binary in addition to kubelet.
+  - An additional component may aggravate the complexity currently. It may be relieved with docker version evolutions.
+- The number of affected users maybe large.
+  - Users must change existing use experience when using Kubernetes and docker.
+  - Users have to change their existing workflows to adapt to this new changes.
+  - And other unrecorded stuff.
+- Updating all the eco-system tools to avoid dependencies on docker.
+- Many people use the built in dockershim for in-cluster image build. While that may not be something we recommend for 
+a variety of reasons, it will be a breaking change for these users.
+- CRI is still in alpha，should probably get a 1.0 out there splitting out dockershim completely from kubelet. There 
+needs to be another KEP for graduating the CRI API version.
+- Existing CNI and CSI plugins may also be affected.
```

Comment:
this surprised me... are there more concrete details you can point to for these?

Reply:
The only CNI-related CRI-vs-dockershim difference that I know of is that CRI handles pod HostPorts itself, while dockershim expects someone else to do it. (Either the plugin does it itself or else the administrator configures a chain of plugins including `portmap`.) So a plugin that previously only supported dockershim might need to be updated to also support CRI correctly (eg, https://github.com/openshift/origin/pull/16463 though that PR was slightly wrong in that it assumed the difference was CRI-O-specific, not CRI in general). But all those issues probably got worked out years ago...

Reply:
Dropping this.